packages/vaex-core/src/superagg.cpp

```diff
@@ -653,6 +653,51 @@ class AggSum : public AggBase<StorageType, StorageType, IndexType> {
     }
 };
 
+template<class StorageType=double, class IndexType=default_index_type, bool FlipEndian=false>
+class AggSumMoment : public AggBase<StorageType, StorageType, IndexType> {
+public:
+    using Base = AggBase<StorageType, StorageType, IndexType>;
+    using Type = AggSumMoment<StorageType, IndexType, FlipEndian>;
+    using Base::Base;
+    AggSumMoment(Grid<IndexType>* grid, uint32_t moment) : Base(grid), moment(moment) {
+    }
+    virtual void reduce(std::vector<Type*> others) {
+        for(auto other: others) {
+            for(size_t i = 0; i < this->grid->length1d; i++) {
+                this->grid_data[i] = this->grid_data[i] + other->grid_data[i];
+            }
+        }
+    }
+    virtual void aggregate(default_index_type* indices1d, size_t length, uint64_t offset) {
+        if(this->data_ptr == nullptr) {
+            throw std::runtime_error("data not set");
+        }
+
+        if(this->data_mask_ptr) {
+            for(size_t j = 0; j < length; j++) {
+                // if not masked
+                if(this->data_mask_ptr[j+offset] == 1) {
+                    StorageType value = this->data_ptr[j+offset];
+                    if(FlipEndian)
+                        value = _to_native(value);
+                    if(value != value) // nan
+                        continue;
+                    this->grid_data[indices1d[j]] += pow(value, moment);
+                }
+            }
+        } else {
+            for(size_t j = 0; j < length; j++) {
+                StorageType value = this->data_ptr[offset + j];
+                if(FlipEndian)
+                    value = _to_native(value);
+                if(value == value) // nan check
```

Comment:
Thank you again for creating and maintaining `vaex`. Hope this is helpful.

This `value == value` condition will be optimised away by most modern compilers when compiling with any optimisation flags, and therefore will not work as expected. One way to check for NANs (at least in `C`) is to add a `volatile` modifier to variable "value". 

I also notice that you are using `pow` for computing the moment, where `moment` is a `size_t`. Might be faster to compute the chained products for `moment = [1, 2, 3] -> [value, value*value, value*value*value]` and then default to `pow` for the general case. 

Thanks!

Reply:
Thank you!

value == value will indeed be optimized away with `--ffast-math`, but I've never seen it been optimized away without it (the unittest will trigger it). But I've never really dug into what is the fasted nan test.

I really like the idea for the moments, I think we should have a moments aggregator, that will do this (up to a particular moment order)! Would you mind opening an issue for this?

cheers,

Maarten

Reply:
Interesting point about the NAN check. All of my reading implies that the use of `volatile` is necessary to ensure correct behaviour. May be that relates to `icc` being over-eager and enabling some degree of `ffast-math` by default. My strategy is always to call `isfinite(float)` and not worry about the run-time performance of the resulting call. (On my OSX laptop, `math.h` does contain two implementations for `isfinite` -> one for the "normal" case, and another for when `ffast-math` is enabled.)

I will open the issue regarding the moments. This is the `C` code I would write for that implementation (likely to be faster than an equivalent `switch` statement):

```C
const StorageType moment_values[] = {1, value, value*value, value*value*value};
const size_t max_moment = sizeof(moment_values)/sizeof(StorageType);
this->grid_data[indices1d[j]] += (moment < max_moment) ? moment_values[moment]: pow(value, moment);
```

Reply:
I will note that my implementation above does not check for overflow/underflow. A better implementation would use an inter-mediate (larger-width) variable to hold the sum, and then cast down to the appropriate type at the end. Plus, depending on how the sums are evaluated, there might need to be a compensated algorithm (e.g, Kahan sum etc)